# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [122]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [123]:
austin = pd.read_csv(r"austin_weather.csv")
austin.head(2)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [124]:
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

In [125]:
austin.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [126]:
pd.set_option("max_columns", 100)
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [127]:
austin["DewPointAvgF"].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [128]:
# there's a string ("-") in the data, therefore Pandas interpret this data as a string

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [129]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [130]:
for col in wrong_type_columns:
    austin[col] = pd.to_numeric(austin[col], errors="coerce")
 

"""
# OR
def austin_conv(col):
    austin[col] = pd.to_numeric(austin[col], errors="coerce")

for columns in wrong_type_columns:
    austin_conv(columns)
"""

'\n# OR\ndef austin_conv(col):\n    austin[col] = pd.to_numeric(austin[col], errors="coerce")\n\nfor columns in wrong_type_columns:\n    austin_conv(columns)\n'

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [131]:
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [132]:
missing_values = austin[austin.isnull().any(axis=1)]
missing_values

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4     2013-12-25         58        50        41           44.0          40.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
42    2014-02-01         76        66        55           62.0          59.0   
51    2014-02-10         60        48        35           49.0          36.0   
...          ...        ...       ...       ...            ...           ...   
1269  2017-06-12         94        85        75           72.0          68.0   
1288  2017-07-01         98        89        79           76.0          72.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1308  2017-07-21        104        91        77           74.0          68.0   
1313  2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
4                   56.0                       30.41   
6                   47.0                       30.46   
7                   36.0                       30.33   
42                  60.0                       29.91   
51                  66.0                       30.23   
...                  ...                         ...   
1269                36.0                       29.96   
1288                38.0                       30.13   
1304                38.0                       29.94   
1308                27.0                       30.01   
1313                25.0                       30.13   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
4                         30.33                      30.27   
6                         30.39                      30.34   
7                         30.17                      30.04   
42                        29.81                      29.75   
51                        30.15                      30.02   
...                         ...                        ...   
1269                      29.89                      29.81   
1288                      30.02                      29.92   
1304                      29.90                      29.84   
1308                      29.95                      29.86   
1313                      30.03                      29.93   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
4                    10.0                10.0                 7.0   
6                    10.0                 9.0                 7.0   
7                    10.0                10.0                 7.0   
42                   10.0                10.0                 9.0   
51                   10.0                 8.0                 4.0   
...                   ...                 ...                 ...   
1269                 10.0                10.0                 2.0   
1288                 10.0                 9.0                 6.0   
1304                 10.0                10.0                10.0   
1308                 10.0                10.0                10.0   
1313                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  Pre

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [133]:
print(austin.shape)
print(missing_values.shape)

# better
print(austin["Date"].size)
print(missing_values["Date"].size)

(1319, 21)
(136, 21)
1319
136


#### Calculate the ratio of missing rows to total rows

In [134]:
col_ratio = missing_values["Date"].size/austin["Date"].size
print(col_ratio)


#col_ratio = missing_values.shape[0]/austin.shape[0]

0.10310841546626232


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [135]:
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [136]:
austin["PrecipitationSumInches"].isna().sum()/austin["PrecipitationSumInches"].size

0.09401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [137]:
austin.drop(columns=["PrecipitationSumInches"], axis=1, inplace=True)
display(austin)

# Print `austin` to confirm the column is indeed removed


Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [138]:
austin_fixed = austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [139]:
display(austin.isna().sum())
display(austin_fixed.isna().sum())

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [140]:
austin_fixed["Events"].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [141]:
austin_fixed[austin_fixed["Events"]=="Fog , Rain , Thunderstorm"]

# OR:
#   - no events
#   - 3 events(Fog, Rain, Thunderstorm)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
114   2014-04-14         73        58        43           69.0          46.0   
142   2014-05-12         82        72        62           76.0          70.0   
157   2014-05-27         73        68        63           67.0          64.0   
185   2014-06-24         93        84        74           74.0          72.0   
186   2014-06-25         89        80        71           76.0          72.0   
208   2014-07-17         89        82        74           76.0          74.0   
257   2014-09-04         99        86        73           77.0          74.0   
271   2014-09-18         90        81        72           75.0          73.0   
336   2014-11-22         67        63        58           64.0          62.0   
438   2015-03-04         70        52        34           67.0          56.0   
483   2015-04-18         77        66        54           66.0          63.0   
500   2015-05-05         83        75        66           70.0          67.0   
508   2015-05-13         70        67        64           70.0          66.0   
518   2015-05-23         87        77        66           74.0          71.0   
520   2015-05-25         82        74        66           74.0          69.0   
540   2015-06-14         86        79        72           76.0          73.0   
542   2015-06-16         83        78        73           75.0          73.0   
543   2015-06-17         89        81        72           77.0          74.0   
545   2015-06-19         84        78        71           76.0          73.0   
546   2015-06-20         83        78        72           77.0          73.0   
547   2015-06-21         86        79        72           78.0          75.0   
556   2015-06-30         94        84        73           76.0          73.0   
628   2015-09-10         96        84        72           76.0          73.0   
678   2015-10-30         75        71        67           73.0          70.0   
848   2016-04-17         78        71        64           71.0          67.0   
861   2016-04-30         86        76        66           74.0          67.0   
889   2016-05-28         91        78        65           77.0          69.0   
949   2016-07-27         92        83        74           78.0          74.0   
969   2016-08-16         83        77        71           76.0          73.0   
1009  2016-09-25         88        79        70           76.0          73.0   
1048  2016-11-03         80        74        68           73.0          70.0   
1123  2017-01-17         59        54        48           55.0          48.0   
1262  2017-06-05         90        78        66           70.0          68.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
114           32.0                 90.0                65.0   
142           61.0                 94.0                87.0   
157           61.0                 93.0                86.0   
185           69.0                 94.0                72.0   
186           68.0                 94.0                77.0   
208           70.0                 94.0                77.0   
257           67.0                 91.0                65.0   
271           70.0                100.0                78.0   
336           56.0                100.0                97.0   
438           32.0                100.0                95.0   
483           53.0                100.0                82.0   
500           62.0                100.0                77.0   
508           64.0                100.0                97.0   
518           66.0                100.0                83.0   
520           66.0                100.0                87.0   
540           71.0                100.0                87.0   
542           70.0                 94.0                84.0   
543           71.0                100.0                83.0   
545           70.0                 94.0                87.0   
546           70.0             

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [142]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

for event in event_list:
    austin_fixed[event]=0


# Print your new dataframe to check whether new columns have been created:
display(austin_fixed)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [143]:
for event in event_list:
    austin_fixed[event] = austin_fixed["Events"].str.contains(event).astype(int)

    
"""
# OR: np.where(condition, output if true, output if false)
for event in event_list:
    austin_fixed[event] = np.where(austin_fixed["Events"].str.contains(event)==True, 1, 0) 
"""

'\n# OR: np.where(condition, output if true, output if false)\nfor event in event_list:\n    austin_fixed[event] = np.where(austin_fixed["Events"].str.contains(event)==True, 1, 0) \n'

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [144]:
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                Events  Snow  Fog  Rain  Thunderstorm  
0  Rain , Thunderstorm     0    0     1             1  
1                          0    0     0             0  
2                          0    0     0             0  
3                          0    0     0             0  
4                          0    0     0             0

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [145]:
austin_fixed.drop(columns=["Events"], axis=1, inplace=True)

In [146]:
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  \
0                 2.0         20.0         4.0         31.0     0    0     1   
1                 5.0         16.0         6.0         25.0     0    0     0   
2                10.0          8.0         3.0         12.0     0    0     0   
3                 7.0         12.0         4.0         20.0     0    0     0   
4                 7.0         10.0         2.0         16.0     0    0     0   

   Thunderstorm  
0             1  
1             0  
2             0  
3             0  
4             0

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [147]:
from datetime import datetime as dt


def convert_date_to_num(date):
    return dt.strptime(str(date), "%Y-%m-%d").toordinal()

austin_fixed["Date"] = austin_fixed["Date"].apply(convert_date_to_num)



# austin_fixed["Date"] = [dt.strptime(date, "%Y-%m-%d").toordinal() for date in austin_fixed["Date"]]

#### Print `austin_fixed` to check your `Date` column.

In [148]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  \
0                 2.0         20.0         4.0         31.0     0    0     1   
1                 5.0         16.0         6.0         25.0     0    0     0   
2                10.0          8.0         3.0         12.0     0    0     0   
3                 7.0         12.0         4.0         20.0     0    0     0   
4                 7.0         10.0         2.0         16.0     0    0     0   

   Thunderstorm  
0             1  
1             0  
2             0  
3             0  
4             0

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [150]:
y = austin_fixed['TempAvgF']
x = austin_fixed[['TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]

# Date is not on the list, so i guess i pass

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [154]:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [161]:
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state = 0, test_size = 0.2)

In [162]:
x_train

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
1261         91        68           71.0          69.0          66.0   
49           65        37           42.0          36.0          29.0   
674          79        58           57.0          54.0          50.0   
1191         86        63           66.0          61.0          51.0   
549          92        75           76.0          73.0          71.0   
...         ...       ...            ...           ...           ...   
763          57        31           32.0          30.0          27.0   
835          83        53           53.0          46.0          38.0   
1216         88        70           68.0          67.0          65.0   
559          91        76           74.0          72.0          69.0   
684          85        72           74.0          72.0          69.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
1261                100.0                76.0                52.0   
49                   85.0                63.0                40.0   
674                  78.0                62.0                45.0   
1191                 93.0                67.0                40.0   
549                  94.0                73.0                52.0   
...                   ...                 ...                 ...   
763                  85.0                61.0                36.0   
835                  66.0                48.0                30.0   
1216                 90.0                73.0                55.0   
559                  88.0                69.0                49.0   
684                 100.0                83.0                65.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
1261                       29.90                      29.83   
49                         30.26                      30.17   
674                        30.05                      29.97   
1191                       29.91                      29.84   
549                        30.18                      30.10   
...                          ...                        ...   
763                        30.37                      30.28   
835                        30.21                      30.15   
1216                       30.07                      30.00   
559                        30.03                      29.96   
684                        30.02                      29.97   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
1261                      29.75                 10.0                 9.0   
49                        30.03                 10.0                 7.0   
674                       29.88                 10.0                10.0   
1191                      29.73                 10.0                 8.0   
549                       30.04                 10.0                10.0   
...                         ...                  ...                 ...   
763                       30.17                 10.0                10.0   
835                       30.10                 10.0                10.0   
1216                      29.91                 10.0                10.0   
559                       29.89                 10.0                 9.0   
684                       29.89                 10.0                 9.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
1261                 2.0         12.0         1.0         24.0     0    0   
49                   3.0         14.0         5.0         21.0     0    0   
674                 10.0         15.0         6.0         23.0     0    0   
1191                 2.0         18.0         8.0         27.0     0    0   
549                 10.0         13.0         5.0         20.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
763                 10.0          8.0         3.0         12.0     0    0   
835                 10.0  

In [163]:
x_test

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
658          90        67           62.0          59.0          58.0   
406          62        49           61.0          51.0          37.0   
1202         79        49           47.0          43.0          40.0   
202          95        71           72.0          67.0          56.0   
1187         83        68           64.0          62.0          58.0   
...         ...       ...            ...           ...           ...   
608          92        72           74.0          71.0          67.0   
194          96        72           71.0          69.0          61.0   
240         102        78           75.0          71.0          64.0   
1154         77        50           60.0          50.0          41.0   
1275        100        78           74.0          71.0          67.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
658                  78.0                56.0                34.0   
406                 100.0                80.0                59.0   
1202                 77.0                52.0                27.0   
202                  90.0                60.0                30.0   
1187                 87.0                67.0                47.0   
...                   ...                 ...                 ...   
608                  93.0                73.0                52.0   
194                  93.0                62.0                30.0   
240                  85.0                59.0                33.0   
1154                 93.0                67.0                41.0   
1275                 85.0                60.0                35.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
658                        30.21                      30.11   
406                        30.26                      30.04   
1202                       30.29                      30.21   
202                        30.15                      30.09   
1187                       30.16                      30.07   
...                          ...                        ...   
608                        30.03                      29.96   
194                        30.13                      30.08   
240                        29.99                      29.90   
1154                       29.98                      29.85   
1275                       29.82                      29.75   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
658                       30.03                 10.0                10.0   
406                       29.90                 10.0                 5.0   
1202                      30.16                 10.0                10.0   
202                       30.03                 10.0                10.0   
1187                      30.00                 10.0                10.0   
...                         ...                  ...                 ...   
608                       29.86                 10.0                 9.0   
194                       30.01                 10.0                10.0   
240                       29.78                 10.0                10.0   
1154                      29.73                 10.0                10.0   
1275                      29.68                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
658                 10.0         10.0         3.0         16.0     0    0   
406                  1.0         15.0         4.0         21.0     0    0   
1202                10.0          9.0         3.0         15.0     0    0   
202                 10.0         13.0         5.0         22.0     0    0   
1187                10.0         14.0         7.0         23.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
608                  4.0         14.0         5.0         23.0     0    0   
194                  6.0  

In [164]:
y_train

1261    80
49      51
674     69
1191    75
549     84
        ..
763     44
835     68
1216    79
559     84
684     79
Name: TempAvgF, Length: 1055, dtype: int64

In [165]:
y_test

658     79
406     56
1202    64
202     83
1187    76
        ..
608     82
194     84
240     90
1154    64
1275    89
Name: TempAvgF, Length: 264, dtype: int64

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [170]:
import math
ts_rows = round(x.shape[0]*0.8)
ts_rows

1055

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [181]:
X_ts_train = x[0:ts_rows]
X_ts_test = x[ts_rows:-1]

display(X_ts_train)
display(X_ts_test)

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0            74        45           67.0          49.0          43.0   
1            56        39           43.0          36.0          28.0   
2            58        32           31.0          27.0          23.0   
3            61        31           36.0          28.0          21.0   
4            58        41           44.0          40.0          36.0   
...         ...       ...            ...           ...           ...   
1050         72        63           68.0          66.0          59.0   
1051         76        59           67.0          60.0          57.0   
1052         71        61           69.0          65.0          61.0   
1053         69        62           66.0          64.0          62.0   
1054         65        60           62.0          61.0          58.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0                    93.0                75.0                57.0   
1                    93.0                68.0                43.0   
2                    76.0                52.0                27.0   
3                    89.0                56.0                22.0   
4                    86.0                71.0                56.0   
...                   ...                 ...                 ...   
1050                100.0                91.0                81.0   
1051                 93.0                85.0                76.0   
1052                100.0                94.0                87.0   
1053                100.0                92.0                84.0   
1054                100.0                94.0                87.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                          29.86                      29.68   
1                          30.41                      30.13   
2                          30.56                      30.49   
3                          30.56                      30.45   
4                          30.41                      30.33   
...                          ...                        ...   
1050                       30.27                      30.22   
1051                       30.19                      30.13   
1052                       30.14                      30.09   
1053                       30.21                      30.16   
1054                       30.29                      30.23   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                         29.59                 10.0                 7.0   
1                         29.87                 10.0                10.0   
2                         30.41                 10.0                10.0   
3                         30.30                 10.0                10.0   
4                         30.27                 10.0                10.0   
...                         ...                  ...                 ...   
1050                      30.14                 10.0                 6.0   
1051                      30.07                 10.0                10.0   
1052                      30.04                 10.0                 5.0   
1053                      30.09                 10.0                 8.0   
1054                      30.18                 10.0                 6.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1050                 1.0          7.0         2.0         12.0     0    0   
1051                 5.0  

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
1055         71        56           58.0          55.0          53.0   
1056         74        55           60.0          57.0          54.0   
1057         77        57           58.0          57.0          54.0   
1058         76        54           58.0          55.0          52.0   
1059         82        56           62.0          58.0          55.0   
...         ...       ...            ...           ...           ...   
1313        103        78           75.0          70.0          61.0   
1314        103        75           71.0          67.0          61.0   
1315        105        76           71.0          64.0          55.0   
1316        107        77           72.0          64.0          55.0   
1317        106        79           70.0          68.0          63.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
1055                 90.0                72.0                53.0   
1056                100.0                79.0                57.0   
1057                 93.0                72.0                50.0   
1058                 93.0                75.0                57.0   
1059                100.0                74.0                48.0   
...                   ...                 ...                 ...   
1313                 91.0                58.0                25.0   
1314                 82.0                54.0                25.0   
1315                 87.0                54.0                20.0   
1316                 82.0                51.0                19.0   
1317                 69.0                48.0                27.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
1055                       30.34                      30.29   
1056                       30.31                      30.25   
1057                       30.28                      30.22   
1058                       30.27                      30.18   
1059                       30.15                      30.09   
...                          ...                        ...   
1313                       30.13                      30.03   
1314                       30.04                      29.97   
1315                       29.97                      29.90   
1316                       29.91                      29.86   
1317                       29.96                      29.91   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
1055                      30.25                 10.0                10.0   
1056                      30.17                 10.0                 8.0   
1057                      30.17                 10.0                 8.0   
1058                      30.12                 10.0                10.0   
1059                      30.00                 10.0                 6.0   
...                         ...                  ...                 ...   
1313                      29.93                 10.0                10.0   
1314                      29.88                 10.0                10.0   
1315                      29.81                 10.0                10.0   
1316                      29.79                 10.0                10.0   
1317                      29.87                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
1055                 8.0          8.0         3.0         12.0     0    0   
1056                 2.0          8.0         2.0         12.0     0    0   
1057                 1.0          9.0         4.0         13.0     0    0   
1058                 9.0          7.0         2.0         12.0     0    0   
1059                 0.0          9.0         2.0         16.0     0    1   
...                  ...          ...         ...          ...   ...  ...   
1313                10.0         15.0         6.0         27.0     0    0   
1314                10.0  

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [182]:
Y_ts_train = y[0:ts_rows]
Y_ts_test = y[ts_rows:-1]

display(Y_ts_train)
display(Y_ts_test)

0       60
1       48
2       45
3       46
4       50
        ..
1050    68
1051    68
1052    66
1053    66
1054    63
Name: TempAvgF, Length: 1055, dtype: int64

1055    64
1056    65
1057    67
1058    65
1059    69
        ..
1313    91
1314    89
1315    91
1316    92
1317    93
Name: TempAvgF, Length: 263, dtype: int64